### PR TITLE
Blacklistfilter: port-filtering

### PR DESCRIPTION
**Pull request by petrmiculek**
Blacklistfilter:

- Added port-filtering (JSON blacklists only):

    - bl_downloader.py parses ports from blacklist, as given by bl_downloader_config.xml
    - Ipblacklistfilter.cpp uses these ports (where available) for more fine-grained detection.

- Extended .blist file format: ip4.blist example
```
IP/prefix,blacklists_bitfield;blacklist:ports;blacklist:ports
192.168.0.0/16,512
192.168.1.18,3;1:8080,5000;2:11111
```
  If no ports are present for given blacklist, its ID only appears in the bitfield value.
  
 - Handling JSON blacklists in bl_downloader also supports filtering entities by the value of a given field (only keep IPs where status=1 (UP) for smashed blacklist)

### CODE REVIEW

File: blacklistfilter/ipblacklistfilter.cpp [hunk 1]
```diff
@@ -292,7 +292,90 @@ int reload_blacklists(black_list_t &v4_list, black_list_t &v6_list, const ip_con
...
 
-            // Add entry to vector
+			// blacklist:[ports] parsing
+			uint16_t bl_num;
+			uint16_t port;
+
+			size_t bl_semicolon_sep = line.find_first_of(';');
+
+			if(bl_semicolon_sep == string::npos)
+			{
```
Comment (havraji6) [inline blacklistfilter/ipblacklistfilter.cpp:302]: Keep the same coding style across the source file:
```
if (...)
{
```
->
```
if (...) {
```

File: blacklistfilter/ipblacklistfilter.cpp [hunk 2]
```diff
@@ -292,7 +292,90 @@ int reload_blacklists(black_list_t &v4_list, black_list_t &v6_list, const ip_con
...
             }
 
-            // Add entry to vector
+			// blacklist:[ports] parsing
+			uint16_t bl_num;
+			uint16_t port;
+
+			size_t bl_semicolon_sep = line.find_first_of(';');
+
+			if(bl_semicolon_sep == string::npos)
```
Comment (havraji6) [inline blacklistfilter/ipblacklistfilter.cpp:301]: `if(` -> `if (`

File: blacklistfilter/ipblacklistfilter.cpp [hunk 3]
```diff
@@ -292,7 +292,79 @@ int reload_blacklists(black_list_t &v4_list, black_list_t &v6_list, const ip_con
                 bl_entry.adaptive_ids = id_part;
             }
 
-            // Add entry to vector
+			// blacklist:[ports] parsing
```
Comment (cejkato2) [inline blacklistfilter/ipblacklistfilter.cpp:295]: Indentation?

File: blacklistfilter/ipblacklistfilter.cpp [hunk 4]
```diff
@@ -292,7 +292,79 @@ int reload_blacklists(black_list_t &v4_list, black_list_t &v6_list, const ip_con
...
+#define state_blacklist_num 1
+#define state_ports 2
+#define state_invalid 3
+#define state_end 4
+
+			int state = state_start;
+			while (state != state_end) {
+				switch (state) {
+					case state_start:
+						if (*index == ';') {
```
Comment (cejkato2) [inline blacklistfilter/ipblacklistfilter.cpp:322]: Coding style

File: blacklistfilter/ipblacklistfilter.cpp [hunk 5]
```diff
@@ -292,7 +292,79 @@ int reload_blacklists(black_list_t &v4_list, black_list_t &v6_list, const ip_con
...
+							index++;
+							state = state_blacklist_num;
+						} else if (*index == '\0') {
+							state = state_end;
+						}
+						else {
```
Comment (cejkato2) [inline blacklistfilter/ipblacklistfilter.cpp:328]: Coding style

File: blacklistfilter/ipblacklistfilter.cpp [hunk 6]
```diff
@@ -292,7 +292,79 @@ int reload_blacklists(black_list_t &v4_list, black_list_t &v6_list, const ip_con
...
+							state = state_invalid;
+						}
+						break;
+
+					case state_blacklist_num:
+						bl_num = strtoul(index, &index, 10);
```
Comment (cejkato2) [inline blacklistfilter/ipblacklistfilter.cpp:334]: Check for errors

File: blacklistfilter/ipblacklistfilter.cpp [hunk 7]
```diff
@@ -292,7 +292,79 @@ int reload_blacklists(black_list_t &v4_list, black_list_t &v6_list, const ip_con
...
+						if (*index == ':') {
+							index++;
+							state = state_ports;
+						} else {
+							state = state_invalid;
+						}
+						break;
+
+					case state_ports:
+						port = strtoul(index, &index, 10);
```
Comment (cejkato2) [inline blacklistfilter/ipblacklistfilter.cpp:346]: Check for errors

File: blacklistfilter/ipblacklistfilter.cpp [hunk 8]
```diff
@@ -292,7 +292,79 @@ int reload_blacklists(black_list_t &v4_list, black_list_t &v6_list, const ip_con
...
+						bl_entry.bl_ports.at(bl_num).insert(port);
```
Comment (cejkato2) [inline blacklistfilter/ipblacklistfilter.cpp:347]: Does bl_port always contain bl_num index?

File: blacklistfilter/ipblacklistfilter.cpp [hunk 9]
```diff
@@ -413,32 +512,56 @@ int blacklist_check(ur_template_t *ur_in,
...
-        } else {
-            ur_set_string(ur_out, detected, F_ADAPTIVE_IDS, "");
-        }
-        return BLACKLISTED;
-    // Check destination IP
+		if (bl[search_result].in_blacklist == ADAPTIVE_BLACKLIST_INDEX) {
+			// Adaptive IP filter mode
+			ur_set_string(ur_out, detected, F_ADAPTIVE_IDS, bl[search_result].adaptive_ids.c_str());
+		} else
+		{
```
Comment (havraji6) [inline blacklistfilter/ipblacklistfilter.cpp:525]:
```suggestion
		} else {
```

File: blacklistfilter/ipblacklistfilter.cpp [hunk 10]
```diff
@@ -413,32 +512,56 @@ int blacklist_check(ur_template_t *ur_in,
...
-        ur_set(ur_out, detected, F_DST_BLACKLIST, 0x0);
-        if (bl[search_result].in_blacklist == ADAPTIVE_BLACKLIST_INDEX) {
-            // Adaptive IP filter mode
-            ur_set_string(ur_out, detected, F_ADAPTIVE_IDS, bl[search_result].adaptive_ids.c_str());
-        } else {
-            ur_set_string(ur_out, detected, F_ADAPTIVE_IDS, "");
-        }
-        return BLACKLISTED;
-    // Check destination IP
+		if (bl[search_result].in_blacklist == ADAPTIVE_BLACKLIST_INDEX) {
```
Comment (cejkato2) [inline blacklistfilter/ipblacklistfilter.cpp:521]: Indentation

File: blacklistfilter/ipblacklistfilter.cpp (hunk 5 above)
Comment (havraji6) [inline blacklistfilter/ipblacklistfilter.cpp:328]:
```suggestion
						} else {
```

File: blacklistfilter/ipblacklistfilter.cpp [hunk 11]
```diff
@@ -292,7 +292,79 @@ int reload_blacklists(black_list_t &v4_list, black_list_t &v6_list, const ip_con
...
 
-            // Add entry to vector
+			// blacklist:[ports] parsing
+			uint16_t bl_num;
+			uint16_t port;
+
+			size_t bl_semicolon_sep = line.find_first_of(';');
+
+			if (bl_semicolon_sep == string::npos)
+			{
```
Comment (havraji6) [inline blacklistfilter/ipblacklistfilter.cpp:302]:
```suggestion
			if (bl_semicolon_sep == string::npos) {
```

**Comment by cejkato2** [comment]
Thank you @petrmiculek for extending features and significant improvement of the source codes!

**Pull request merged**